client/mm/libxc/binance.go

```diff
@@ -0,0 +1,1168 @@
+
+// Connect connects to the binance API.
+func (bnc *binance) Connect(ctx context.Context) error {
+	bnc.ctx = ctx
+
+	if err := bnc.getCoinInfo(); err != nil {
+		return fmt.Errorf("error getting coin info: %v", err)
+	}
+
+	if err := bnc.getMarkets(); err != nil {
+		return fmt.Errorf("error getting markets: %v", err)
+	}
+
+	if err := bnc.getUserDataStream(); err != nil {
+		return err
+	}
+
+	// Refresh the markets periodically.
+	go func() {
+		nextTick := time.After(time.Hour)
+		for {
+			select {
+			case <-nextTick:
+				err := bnc.getMarkets()
+				if err != nil {
+					bnc.log.Errorf("error fetching markets: %v", err)
+					nextTick = time.After(time.Minute)
+				} else {
+					nextTick = time.After(time.Hour)
+					bnc.sendCexUpdateNotes()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// Refresh the coin info periodically.
+	go func() {
+		nextTick := time.After(time.Hour)
+		for {
+			select {
+			case <-nextTick:
+				err := bnc.getCoinInfo()
+				if err != nil {
+					bnc.log.Errorf("error fetching markets: %v", err)
+					nextTick = time.After(time.Minute)
+				} else {
+					nextTick = time.After(time.Hour)
+					bnc.sendCexUpdateNotes()
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// SubscribeCEXUpdates returns a channel which sends an empty struct when
+// the balance of an asset on the CEX has been updated.
+func (bnc *binance) SubscribeCEXUpdates() <-chan interface{} {
+	updater := make(chan interface{}, 128)
+	bnc.cexUpdatersMtx.Lock()
+	bnc.cexUpdaters = append(bnc.cexUpdaters, updater)
+	bnc.cexUpdatersMtx.Unlock()
+	return updater
+}
+
+// Balance returns the balance of an asset at the CEX.
+func (bnc *binance) Balance(symbol string) (*ExchangeBalance, error) {
+	assetConfig, err := bncSymbolData(symbol)
+	if err != nil {
+		return nil, err
+	}
+
+	bnc.balanceMtx.RLock()
+	defer bnc.balanceMtx.RUnlock()
+
+	bal, found := bnc.balances[assetConfig.coin]
+	if !found {
+		return nil, fmt.Errorf("no %q balance found", assetConfig.coin)
+	}
+
+	return &ExchangeBalance{
+		Available: uint64(bal.available * float64(assetConfig.conversionFactor)),
+		Locked:    uint64(bal.locked * float64(assetConfig.conversionFactor)),
+	}, nil
+}
+
+// GenerateTradeID returns a trade ID that must be passed as an argument
+// when calling Trade. This ID will be used to identify updates to the
+// trade. It is necessary to pre-generate this because updates to the
+// trade may arrive before the Trade function returns.
+func (bnc *binance) GenerateTradeID() string {
+	nonce := bnc.tradeIDNonce.Add(1)
+	nonceB := encode.Uint32Bytes(nonce)
+	return fmt.Sprintf("%x", append(encode.RandomBytes(10), nonceB...))
+}
+
+// Trade executes a trade on the CEX. updaterID takes an ID returned from
+// SubscribeTradeUpdates, and tradeID takes an ID returned from
+// GenerateTradeID.
+func (bnc *binance) Trade(baseSymbol, quoteSymbol string, sell bool, rate, qty uint64, updaterID int, orderID string) error {
```

Comment:
Comment says tradeID but there is no argument by that name.

Should this check that orderID is a valid order ID? Or should the variable be called tradeID? Then should it check it is the correct length? Should it check that it is unique in the `bnc.tradeToUpdater` map?